### PR TITLE
Add linter for unitless durations

### PR DESCRIPTION
**Pull request by josharian**

### CODE REVIEW

File: lint.go [hunk 1]
```diff
@@ -1394,6 +1396,59 @@ func (f *file) lintTimeNames() {
...
+	f.walk(func(node ast.Node) bool {
+		e, ok := node.(ast.Expr)
+		if !ok {
+			return true
+		}
+
+		// Don't complain about time.Duration(30), at least for now.
+		if call, ok := node.(*ast.CallExpr); ok {
+			if len(call.Args) == 1 {
+				sel, ok := call.Fun.(*ast.SelectorExpr)
```
Comment (dsymonds) [inline lint.go]: use the isPkgDot helper func.

File: lint.go [hunk 2]
```diff
@@ -1394,6 +1396,59 @@ func (f *file) lintTimeNames() {
...
+		// Ok for 0 to be unitless; non-constants are ok
+		if x, ok := exact.Int64Val(tv.Value); !ok || x == 0 {
+			return true
+		}
+		if _, ok := e.(*ast.BasicLit); !ok {
+			// Don't descend into complex expressions
+			return false
+		}
+
+		f.errorf(e, 0.8, category("time"), "time.Duration without unit: %v", f.render(e))
```
Comment (dsymonds) [inline lint.go]: I think the message needs to be something like "time.Duration expression should use time package constants".

File: testdata/time.go [hunk 3]
```diff
@@ -11,3 +11,27 @@ import (
 var rpcTimeoutMsec = flag.Duration("rpc_timeout", 100*time.Millisecond, "some flag") // MATCH /Msec.*\*time.Duration/
 
 var timeoutSecs = 5 * time.Second // MATCH /Secs.*time.Duration/
+
+var delay = flag.Duration("delay", 5, "wait impatiently") // MATCH /time\.Duration without unit: 5/
+
+func f(...time.Duration) {}
+
+var (
+	_ = time.Duration(15)
```
Comment (dsymonds) [inline testdata/time.go]: why isn't this getting flagged?

**Comment by dsymonds** [comment]
Have you checked that this behaves gracefully on code that doesn't fully type check (e.g. because an import can't be found)?

File: lint.go (hunk 1 above)
Comment (josharian) [inline lint.go]: Deleted instead.

File: lint.go (hunk 2 above)
Comment (josharian) [inline lint.go]: Done. Used some `.*`s in the tests, since it was long. I'll use the full text there if you prefer.

File: testdata/time.go (hunk 3 above)
Comment (josharian) [inline testdata/time.go]: I saw a lot of `time.Duration(30) * time.Second` when I was flagging it, which is maybe not ideal, but seemed intentional and not the end of the world. It also would need special handling to catch. I'll add that handling if you would like--let me know.

**Comment by josharian** [comment]
I have confirmed. It is also pretty clear from the code--when input doesn't fully type check, the type/value lookup `f.pkg.typesInfo.Types[e]` yields nils, and the check stops immediately in that case.

**Comment by josharian** [comment]
I'm not sure whether you squash or I do (I miss gerrit), so I added another commit. Happy to squash if you prefer. lmk

File: testdata/time.go (hunk 3 above)
Comment (dsymonds) [inline testdata/time.go]: Okay, the message probably isn't great for `time.Duration(30) * time.Second` either, even though we'd typically ask people to change it. I'm fine with not catching this (at least right now), but please add a comment to the test data file saying why.

File: lint.go [hunk 4]
```diff
@@ -1394,6 +1396,47 @@ func (f *file) lintTimeNames() {
...
 
+func (f *file) lintDurationUnits() {
+	f.walk(func(node ast.Node) bool {
+		e, ok := node.(ast.Expr)
+		if !ok {
+			return true
+		}
+
+		tv := f.pkg.typesInfo.Types[e]
+		if tv.Type == nil || !f.pkg.isNamedType(tv.Type, "time", "Duration") {
```
Comment (dsymonds) [inline lint.go]: I don't think you need the nil check here

File: lint.go [hunk 5]
```diff
@@ -1394,6 +1396,47 @@ func (f *file) lintTimeNames() {
...
+			return true
+		}
+		if tv.Value == nil {
+			return true
+		}
+		// Ok for 0 to be unitless; non-constants are ok
```
Comment (dsymonds) [inline lint.go]: Ok -> OK

(that was confusing for me to read right now; it looked like `0k`)

File: lint.go [hunk 6]
```diff
@@ -1394,6 +1396,47 @@ func (f *file) lintTimeNames() {
...
+		if x, ok := exact.Int64Val(tv.Value); !ok || x == 0 {
+			return true
+		}
+		if _, ok := e.(*ast.BasicLit); !ok {
```
Comment (dsymonds) [inline lint.go]: hoist this up higher? In fact, why not make this the check up on line 1401?

**Comment by dsymonds** [comment]
Yes, please squash, though I'll end up cherry-picking when this is ready for submitting anyway. Rebase on the latest tip too; a few other changes have come in since you started.

File: lint.go (hunk 4 above)
Comment (josharian) [inline lint.go]: Done. Thanks.

File: lint.go (hunk 5 above)
Comment (josharian) [inline lint.go]: Done

File: lint.go (hunk 6 above)
Comment (josharian) [inline lint.go]: It is important that the earlier checks return true, whereas this returns false. I've expanded this a bit and given it a (hopefully helpful) comment. In the process, I also added handling of `time.Duration(30)`.

File: testdata/time.go (hunk 3 above)
Comment (josharian) [inline testdata/time.go]: `time.Duration(30)` now gets flagged. There are some other cases that get missed; there is a discussion in the comment in the newest commit.

**Comment by josharian** [comment]
Updated, rebased and squashed. Thanks for all the useful feedback. PTAL.

File: lint.go [hunk 7]
```diff
@@ -1396,6 +1398,70 @@ func (f *file) lintTimeNames() {
...
+		if tv.Value == nil {
+			return true
+		}
+		// OK for 0 to be unitless; non-constants are OK.
+		if x, ok := exact.Int64Val(tv.Value); !ok || x == 0 {
+			return true
+		}
+
+		// e has a constant value. What matters now is how it was defined.
+		// It definition might be spread across many expressions:
```
Comment (dsymonds) [inline lint.go]: Its

File: testdata/time.go [hunk 8]
```diff
@@ -11,3 +11,31 @@ import (
...
+const x = 10
+
+func g() {
+	time.Sleep(
+		12, // MATCH /time.*expression.*constants: 12/
+	)
+	time.Sleep(x) // Not caught; see comments in lintDurationUnits
+	_ = f(1,      // MATCH /time.*expression.*constants: 1$/
+		1e12) // MATCH /time.*expression.*constants: 1e12$/
+}
```
Comment (dsymonds) [inline testdata/time.go]: verify somewhere that `0` is okay.

File: lint.go [hunk 9]
```diff
@@ -1396,6 +1398,70 @@ func (f *file) lintTimeNames() {
...
+			// time.Sleep(x) where 'const x = 10'.
+			return false
+		case *ast.CallExpr:
+			// Descend into call expressions in order to catch time.Duration(15).
+			return true
+		case *ast.BasicLit:
+			// A basic literal cannot possibly use time package constants.
+		}
+
+		f.errorf(e, 0.8, category("time"), "time.Duration expression should use time package constants: %v", f.render(e))
```
Comment (dsymonds) [inline lint.go]: `time.Duration expression "%v" should use time package constants`

it reads a bit more naturally, and will make the test a little clearer

**Comment by dsymonds** [comment]
Thanks. This is almost there.

**Comment by josharian** [comment]
All done. PTAL

> Thanks. This is almost there.

I consider it a privilege to get code reviews. :)

**Comment by dsymonds** [comment]
LGTM

**Comment by dsymonds** [comment]
Merged in b550591.

**Pull request closed without merging**